Cython/Compiler/ExprNodes.py

```diff
@@ -6158,6 +6164,77 @@ def set_py_result_type(self, function, func_type=None):
+            if len(args) != 2:
+                error(self.args.pos, "function_type requires exactly two positional arguments.")
+                return None
+            func_args, func_return = args
+            if not isinstance(func_args, ListNode):
+                error(func_args.pos, "First argument of function_type must be a list literal with parameter types.")
+                return None
+            if kwargs is not None and not isinstance(kwargs, DictNode):
+                error(kwargs.pos, "function_type kwargs should be in a dictionary.")
+                return None
+            param_types = [arg.analyse_as_type(env) for arg in func_args.args]
+            if None in param_types:
+                error(func_args.args[param_types.index(None)].pos, "Unknown type in parameter types")
+                return None
+            ret_type = func_return.analyse_as_type(env)
+            if ret_type is None:
+                error(func_return.pos, "Unknown return type")
+            exc_value = None
+            noexcept = False
+            exc_check = False
+            exc_clause = False
+            func_type_kwargs = {}
+            if kwargs is not None:
+                for kv_pair in kwargs.key_value_pairs:
+                    k = kv_pair.key.constant_result
+                    v = kv_pair.value.constant_result
+                    if k in ('nogil', 'has_varargs', 'noexcept', 'check_exception'):
+                        if isinstance(v, bool):
+                            if k == 'noexcept':
+                                noexcept = v
+                            elif k == 'check_exception':
+                                exc_clause = True
+                                exc_check = v
+                            else:
+                                func_type_kwargs[k] = v
+                        else:
+                            error(self.pos, f"Value for {k} must be boolean")
+                    elif k == 'exceptval':
+                        exc_value = v
+                        exc_clause = True
+                    elif k == 'except_plus':
+                        exc_clause = True
+                        if v == '*':
+                            exc_check = '+'
+                            exc_value = CharNode(kv_pair.value.pos, value='*')
+                        elif isinstance(v, str):
+                            exc_check = '+'
+                            exc_value = NameNode(kv_pair.value.pos, value=v)
+                        elif isinstance(v, bool):
+                            exc_check = '+' if v else False
+                        else:
+                            error(self.pos, f"Value for {k} must be '*', string or boolean, was {v!r}")
+
+                    else:
+                        error(self.pos, f"Unknown kwarg in function_type: {k}")
+            if noexcept and exc_clause:
+                error(self.pos, "Cannot combine noexcept=True with another exception clause.")
```

Comment:
We could provide a better error position here if we detected this case inside of the loop rather than afterwards.